File: performance-comparison/benchmarks/io/python/random_access.py
import sys, os

file_size = int(sys.argv[1]) if len(sys.argv) > 1 else 1048576
n_reads   = int(sys.argv[2]) if len(sys.argv) > 2 else 1000
path      = "/tmp/bench_io_random_py.bin"
MASK      = (1 << 64) - 1

with open(path, 'wb') as fw:
    pos = 0
    while pos < file_size:
        chunk = min(4096, file_size - pos)
        fw.write(bytes((pos + i) & 0xFF for i in range(chunk)))
        pos += chunk

state, checksum = 12345678, 0
with open(path, 'rb') as fr:
    for _ in range(n_reads):
        state = (state * 6364136223846793005 + 1442695040888963407) & MASK
        fr.seek((state >> 1) % file_size)
        b = fr.read(1)
        if b: checksum += b[0]
os.remove(path)
print(checksum)

File: performance-comparison/benchmarks/io/clojure/random_access.clj
(let [file-size (Long/parseLong (first *command-line-args*))
     n-reads    (Long/parseLong (second *command-line-args*))
     path       "/tmp/bench_io_random_clj.bin"]
  ;; write deterministic data
  (with-open [fw (java.io.RandomAccessFile. path "rw")]
    (.setLength fw file-size)
    (loop [pos 0 seq-v 0]
      (when (< pos file-size)
        (let [chunk (int (min 4096 (- file-size pos)))
              buf   (byte-array chunk (map unchecked-byte (range seq-v (+ seq-v chunk))))]
          (.seek fw pos)
          (.write fw buf)
          (recur (+ pos chunk) (+ seq-v chunk))))))
  ;; random reads
  (let [raf (java.io.RandomAccessFile. path "r")
        checksum (loop [i 0 state (long 12345678) cs (long 0)]
                   (if (>= i n-reads) cs
                     (let [state2 (unchecked-add (unchecked-multiply state 6364136223846793005) 1442695040888963407)
                           offset (Math/abs (rem (unsigned-bit-shift-right state2 1) file-size))]
                       (.seek raf offset)
                       (recur (inc i) state2 (+ cs (.read raf))))))]
    (.close raf)
    (.delete (java.io.File. path))
    (println checksum)))
